README.md
# egg_tandem_hunter v1.0.0
This app has been adapted in house to work for samples aligned to b38 reference.

## Requirements
-   Python 2.7+/3.4+ (originally tested on python 2.7 and python 3.6)
-   pandas
-   numpy

A requirements file for based on python 3.6 is available in the `requirements.txt` file.

## Purpose
Identification of duplicated/amplified regions based on the comparison of normalise coverage between two intervals in given coverage file.

## Usage
TandemHunter.py assesses duplicated/amplified regions based on the comparison of
normalised coverage between two intervals in given coverage file. Intervals are given in
.json format (tandemhunter-main/intervals_b38.json)

intervals_b38.json can be used to identify partial tandem duplications and is based on
the method described by McKerrell et al., 2016, where they compare
the normalised coverage of exon 3 and exon 27 of MLL/KMT2A.

The coverage file should be in the following tab-delimited format, which matches the PER_TARGET_COVERAGE file generated by Picard CollectHsMetrics:

```
chrom	start	end	length	name	%gc	mean_coverage	normalized_coverage	min_normalized_coverage	max_normalized_coverage	min_coverage	max_coverage	pct_0x	read_count
chr11	118436493	118436944	452	.	0.792035	2120.834071	0.750534	0.4735	0.952308	1338	2691	0	10154
chr11	118468776	118468844	69	.	0.463768	5666.768116	2.005392	1.758461	2.115179	4969	5977	0	8063
chr11	118471663	118474315	2653	.	0.455711	7069.555974	2.501819	1.547899	2.771638	4374	7832	0	164434
```

NOTE: The exact coordinates for exons 3 and 27 need to be in the
PER_TARGET_COVERAGE output file, this can be done by running Picard using
egg_tandem_hunter/Picard_inputs/KMT2A_all_exons.bed. This file has the exonic
coordinates for KMT2A and used to retrieve coverage information for each exon.

TandemHunter can be run with a single PER_TARGET_COVERAGE file (with `-F` or `--file` switch) or a batch of
coverage files (`-B` or `--batch` switch) as shown below. One of these arguments has to be given, but not both.

```
# Pass in a batch directory and write to specified directory
python TandemHunter.py -B test/Batch -O /path/to/output_dir

# Pass in a single coverage file and write to specified directory
python TandemHunter.py -F test/PositiveSample.qc.coverage.txt -O /path/to/output_dir

```

A full list of additional arguments can be viewed by `python TandemHunter.py --help`.

Coverage comparisons are written to a files called `.cvg_comparison.csv` for each input file in the present working directory by default.

An example of the contents of a `.cvg_comparison.csv` file is given below.

```
sample_id,MLL_EXON3,MLL_EXON27,fold_change,log2_fold_change,above_cut_off
2204845_22144Z0041_1_BM_MPD_MYE_M_EGG2_S39_L001_markdup.pertarget_,2.501819,0.9739260000000001,2.568797835,1.361093354,TRUE

```

generate_comparison_csv_to_xls.py
generate_comparison_csv_to_xls.py takes outputs from tandemhunter to
make a single .xlsx spreadsheet named comparison_csv.xlsx

An example of it's contents comparison_csv.xlsx is given below.

```
sample_id	MLL_EXON3	MLL_EXON27	fold_change	log2_fold_change	above_cut_off
sample_1	1.469901	1.470338	0.999702789	-0.000428849	False
sample_2	1.452241	1.476602	0.983501986	-0.024000129	False
sample_3	1.444375	1.471847	0.981335016	-0.027182356	False
```
## Testing
The following non-base module(s) are required for running the unit tests. These
are defined in the `requirements.txt` file.
-   pytest
-   pytest-cov

To run the unit tests and assess code coverage, run the following command.

```bash
# run this line in the base directory for the module
pytest --cov=TandemHunter --cov-report term-missing -v test
```

Originally developed by NEY GLH Bioinformatics Team.


tandemhunter-main/ChangeLog.md
# ChangeLog

## v2.0.0
- Working version.

# Changes
-Tool modified (.json) to work for b38 coordinates for KMT2A exons
-Tool works for ".pertarget_coverage.tsv" from eggd_picard output
-Additional script added to process tool outputs into single .xlsx file
-Renamed original directory
-Updated TandemHunter readme.md


tandemhunter-main/test/test_TandemHunter.py
# Unit tests for Tandem Hunter
# Author: Helena L. Spiewak
# Date created: 2021-04-07
# Date modified:

from ..TandemHunter import TDHunter, tdh_argument_parser
import pytest
import os
import csv
import glob


class BaseTest(object):
    intervals = "./test/test_intervals.json"
    out_dir = "./test/test_output"
    cov_file_pattern = ".qc.coverage.txt"
    out_fname_suffix = "*.cvg_comparison.csv"

    def teardown_method(self, method):
        """ teardown any state that was previously setup with a setup_method
        call.
        """
        # find breakpoints results files
        files = glob.glob(
            os.path.join(self.out_dir, self.out_fname_suffix), recursive=True,
        )
        # delete them
        [os.remove(f) for f in files]


class TestTDHunterAll(BaseTest):
    def check_file(self, output_file, output_contents):
        """ Check results files.
        :param: output_file: Output file to check
        :param: output_contents: Expected output contents
        :returns: Asserts each line is true
        """
        # check file created
        assert os.path.isfile(output_file)

        # open file and check candidate breakpoints
        with open(output_file, "r") as results:
            results_reader = csv.reader(results)
            n = 0
            for row in results_reader:
                assert row == output_contents[n]
                n += 1
            # check number of breakpoints as expected
            assert n == len(output_contents)

    def test_with_positive_file(self):
        """ No error if file argument given and results file generated that
        contains positive results """
        file = "./test/PositiveSample.qc.coverage.txt"
        # initiate class
        td = TDHunter(file=file, intervals=self.intervals, out_dir=self.out_dir)
        assert td

        # check and contents
        self.check_file(
            output_file=td.fname,
            output_contents=[
                [
                    "sample_id",
                    "MLL_EXON3",
                    "MLL_EXON27",
                    "fold_change",
                    "log2_fold_change",
                    "above_cut_off",
                ],
                [
                    "PositiveSample",
                    "2.284132",
                    "1.250261",
                    "1.826924138",
                    "0.869416728",
                    "TRUE",
                ],
            ],
        )

    def test_with_negative_file(self):
        """ No error if file argument given and results file generated that
        contains negative results """
        file = "./test/NegativeSample.qc.coverage.txt"
        # initiate class
        td = TDHunter(file=file, intervals=self.intervals, out_dir=self.out_dir)
        assert td

        # check and contents
        self.check_file(
            output_file=td.fname,
            output_contents=[
                [
                    "sample_id",
                    "MLL_EXON3",
                    "MLL_EXON27",
                    "fold_change",
                    "log2_fold_change",
                    "above_cut_off",
                ],
                [
                    "NegativeSample",
                    "1.237117",
                    "1.187973",
                    "1.041367944",
                    "0.058479903",
                    "FALSE",
                ],
            ],
        )

    def test_with_interval_json_no_dup_threshold(self):
        """ No error if interval json file does not have dup_threshold defined. """
        file = "./test/PositiveSample.qc.coverage.txt"
        intervals = "./test/test_intervals_no_dup_threshold.json"
        # initiate class
        td = TDHunter(file=file, intervals=intervals, out_dir=self.out_dir)
        assert td

        # check and contents
        self.check_file(
            output_file=td.fname,
            output_contents=[
                [
                    "sample_id",
                    "MLL_EXON3",
                    "MLL_EXON27",
                    "fold_change",
                    "log2_fold_change",
                    "above_cut_off",
                ],
                [
                    "PositiveSample",
                    "2.284132",
                    "1.250261",
                    "1.826924138",
                    "0.869416728",
                    "TRUE",
                ],
            ],
        )

    def test_initiation_with_batch(self):
        """ No error if batch argument given """
        td = TDHunter(
            batch="./test/Batch", out_dir=self.out_dir, intervals=self.intervals
        )
        assert td
        # check 2 results file created
        results_files = glob.glob(
            os.path.join(self.out_dir, self.out_fname_suffix), recursive=True
        )
        assert len(results_files) == 2

    def test_system_exit_no_file_batch(self):
        """ Initiation with no file or batch argument causing system exit """
        with pytest.raises(SystemExit):
            TDHunter()

    def check_invalid_file(self, capsys, file, expected_stderr):
        """ Method to check invalid file.
        :param: file: Path to test file
        :param: expected_stderr: Expected stderr content
        :returns: Asserts file doesn't exists and stderr message
        """
        # initiate class
        td = TDHunter(file=file, intervals=self.intervals, out_dir=self.out_dir)
        assert td
        # checkout std err message in output
        captured = capsys.readouterr()
        for err in expected_stderr:
            assert err in captured.err
        # check that no file was created
        try:
            td.fname
        except AttributeError:
            assert True

    def test_with_bad_header_file(self, capsys):
        """ Error if coverage file has bad header """
        file = "./test/BadHeader.qc.coverage.txt"
        self.check_invalid_file(
            capsys=capsys,
            file=file,
            expected_stderr=[
                "cannot safely convert passed user dtype",
                "coverage file is invalid",
            ],
        )

    def test_with_missing_intervals_file(self, capsys):
        """ Error if file does not contain the given intervals """
        file = "./test/MissingIntervals.qc.coverage.txt"
        self.check_invalid_file(
            capsys=capsys,
            file=file,
            expected_stderr=[
                "No objects to concat",
                "coverage file is invalid",
            ],
        )

    def test_with_malformed_intervals_file(self, capsys):
        """ SystemExit if interval file not as expected """
        file = "./test/MissingIntervals.qc.coverage.txt"
        intervals = "./test/test_intervals_malformed.json"
        with pytest.raises(SystemExit):
            TDHunter(file=file, intervals=intervals)

class TestFindCoverageFiles(BaseTest):
    def test_coverage_file_present(self):
        """ Returns a list of paths to coverage file if given correct
        directory containing expected coverage files """
        coverage_files = TDHunter._find_coverage_files(
            batch_dir="./test/Batch", cov_file_pattern=self.cov_file_pattern
        )
        # returns a list of length 1
        assert isinstance(coverage_files, list)
        assert len(coverage_files) == 2

    def test_coverage_file_not_present(self):
        coverage_files = TDHunter._find_coverage_files(
            batch_dir="./test/EmptyBatch",
            cov_file_pattern=self.cov_file_pattern,
        )
        # returns a list of length 0
        assert isinstance(coverage_files, list)
        assert len(coverage_files) == 0


class TestTDHArgsParser(BaseTest):
    def test_parser_file(self):
        """ File argument should be as expected """
        file = "./test/PositiveSample.qc.coverage.txt"
        parser = tdh_argument_parser(
            ["-F", file, "--intervals", self.intervals]
        )
        assert parser.file == file

    def test_parser_batch(self):
        """ Batch argument should be as expected """
        batch = "./test/"
        parser = tdh_argument_parser(
            ["-B", batch, "--intervals", self.intervals]
        )
        assert parser.batch == batch


src/code.sh
#!/bin/bash

# Runs TandemHunter to produce individual comparison_csv files with PTD predictions
# Also generates single .xlsx spreadsheet with PTD predictions for query
# per_target_coverage files


# -e = exit on error; -x = output each line that is executed to log; -o pipefail = throw an error if there's an error in pipeline
set -e -x -o pipefail

# Make directories to hold outputs
mkdir /home/dnanexus/out
mkdir /home/dnanexus/out/summary_csvs
mkdir /home/dnanexus/out/summary_xlsx

# Download inputs from DNAnexus
dx-download-all-inputs

# handle when NO query coverage files provided
if [ -z "$coverage_targets_tsv" ]; then
	touch /home/dnanexus/out/summary_csv/noQuerycoveragefileFound.summary_csv
else
    # Download inputs from DNAnexus
    dx-download-all-inputs

# run script for PTD prediction outputs
python TandemHunter.py $advanced_options "$output_name" "${coverage_targets_tsv[@]}"

# Run second script (if number of files greater than 1)
# collate data into single dataframe
if f in ["$coverage_targets_file"]!= 0; then
    python TandemHunter.py $advanced_options $run "$output_name" "${comparison_csv[@]}"
else
    mv ~/"${comparison_csv_prefix}.pertarget_coverage.tsv" ~/out/summary_csvs/
    dx-upload-all-outputs --parallel

# Upload outputs (from /home/dnanexus/out) to DNAnexus
dx-upload-all-outputs --parallel

resources/tandemhunter-main/generate_comparison_csv_to_xls.py
"""
script to generate .xls file from a run of cvg_comparison.csv
generated from TandemHunter

"""

import argparse
from pathlib import Path
import pty
import subprocess
import os
import sys
import pytest

import pandas as pd


def parse_args():
    """Parse command line arguments"""
    parser = argparse.ArgumentParser()

    parser.add_argument(
        '--comparison_csv', type=argparse.FileType('r'), nargs="+",
        required=True,
        help='all cvg_comparison.csv files for tandemhunter output'
    )
    parser.add_argument(
        '--run',
        help='name of run to prefix output .xls file with'
    )

    args = parser.parse_args()
    # print(args)
    # print(args.comparison_csv)

    return args


def generate_PTD_KMT2A_df(args):
    """Generates data frame from multiple per cvg_comparison.csv
    from tandemhunter

    Returns:
    -PTD_KMT2A(df):df of normalized coverages for KMT2A exons 3 and 27,
    fold coverage difference between exons and algorithm predictions
    for PTDs in KMT2A (as either true/false)
    """
    # create empty list for data frames
    all_data = []
    for index, file in enumerate(args.comparison_csv):
        df = pd.read_csv(file.name, delimiter=',')
        # append rows
        all_data.append(df)
    # merge rows into single data frame
    merged_data = pd.concat(all_data)
    print(merged_data)

    # check df has no duplicate results
    assert merged_data['sample_id'].is_unique, 'df has duplicates'

    return merged_data


def KM2TA_df_to_xls(PTD_KM2TA_1, args):
    """"Generates .xls spreadsheet with PTD predictions from TandemHunter

   Returns:
   .xls file with fold coverage differences for
   normalized coverage for exons 3 and 27 of KM2TA gene and
   TRUE/FALSE statement to suggest if sample likely has PTDs
   """

    # print(PTD_KM2TA_1.dtypes)
    # convert boolean to string to keep true/false
    PTD_KM2TA_1['above_cut_off'] = PTD_KM2TA_1['above_cut_off'].astype('str')
    #write to excel
    if (args.run):
        PTD_KM2TA_1.to_excel((args.run) + "_comparison_csv.xlsx",
        index=False, header=True)
    else:
        PTD_KM2TA_1.to_excel("comparison_csv.xlsx", index=False, header=True)


def main():
    args = parse_args()
    PTD_KM2TA_1 = generate_PTD_KMT2A_df(args)

    # write output bed file
    KM2TA_df_to_xls(PTD_KM2TA_1, args)


if __name__ == "__main__":
    main()


tandemhunter-main/TandemHunter.py
# Tandem Hunter: Identification of duplicated/amplified
# regions based on the comparison
# of normalise coverage between two intervals in given coverage file
# (PER_TARGET_COVERAGE from Picard CollectHsMetrics).

# Author: Helena L. Spiewak
# Contributors: Nadia Mohammed
# Date created: 2021-04-06
# Date modified: 2021-04-06
# Date modified for v2.0.0: 2022-08-18

# It is used to identify KMT2A/MLL partial tandem duplications and is based on
# the method described by McKerrell et al., 2016, where they compare
# the coverage of exon 3 and exon 27 of MLL/KMT2A.

import pandas as pd
import numpy as np
import argparse
import sys
import os
import json
import multiprocessing as mp
import fnmatch
import functools
from datetime import datetime
import glob

# define Picard CollectHsMetrics PER_TARGET_COVERAGE columns
COV_COLUMN_DTYPES = {
    "chrom": "category",
    "start": "int",
    "end": "int",
    "length": "int",
    "name": "category",
    "%gc": "float64",
    "mean_coverage": "float64",
    "normalized_coverage": "float64",
    "min_normalized_coverage": "float64",
    "max_normalized_coverage": "float64",
    "min_coverage": "int",
    "max_coverage": "int",
    "pct_0x": "float64",
    "read_count": "int",
}


def tdh_argument_parser(args):
    """ Parse arguments for Tandem Hunter. """
    parser = argparse.ArgumentParser(
        description=(
            "Tandem Hunter: Identification of duplicated/amplified regions"
            "based on comparing normalised coverage between two intervals"
        )
    )
    required = parser.add_mutually_exclusive_group(required=True)
    required.add_argument(
        "-F", "--file", default=False, help="Path to coverage file OR"
    )
    required.add_argument(
        "-B",
        "--batch",
        default=False,
        help="Path to coverage files folder. DON'T use -F and -B together.",
    )
    parser.add_argument(
        "--intervals",
        type=str,
        required=True,
        help=(
            "Path to json file containing intervals to compare coverage at."
            " These regions have to be in the coverage file as is as this "
            " programme uses pattern matching"
        ),
    )
    parser.add_argument(
        "--dup-threshold",
        type=float,
        default=1.122995,
        help="Default threshold for dup/amp regions if not in interval file",
    )
    parser.add_argument(
        "-O", "--out_dir",
        default=".", help="Output write directory"
    )
    parser.add_argument(
        "--out-fname-suffix",
        default=".cvg_comparison.csv",
        help="Filename suffix to use when naming results file",
    )
    parser.add_argument(
        "--processes", default=0,
        help="Number of processes to run in parallel",
    )
    parser.add_argument(
        "--cov-file-pattern",
        default="coverage.tsv",
        help="Filename suffix used to identify coverage files in batch mode",
    )
    parser.add_argument(
        "--metric",
        default="normalized_coverage",
        help="Metric to used in comparison",
    )
    parser.add_argument("--version",
        action="version",
        version="%(prog)s 2.0.0")

    return parser.parse_args(args)


class TDHunter(object):
    """ Object containg methods that compare normalised coverage
        at two given intervals.
        :param: file: Path to coverage file
        :param: batch: Path to folder with a batch of coverage files
        :param: intervals: Path to file with intervals to compare coverage at
        :param: dup_threshold: Threshold used to identify dup/amp regions
        :param: out_dir: Output write directory
        :param: out_fname_suffix: Filename suffix for naming results file
        :param: processes: Maximum number of processes to run (default is cpu
                            count)
        :param: cov_file_pattern: Filename suffix to identify coverage files
                                    when in batch mode
        :param: column_dtypes: Dictionary with column name and datatype for
                            the given coverage file type
        :param: metric: Metric to used in comparison
        :returns: Writes a coverage comparison between two intervals to a text
                    file on per-sample basis
    """
    def __init__(
        self,
        file=False,
        batch=False,
        intervals=False,
        dup_threshold=1.122995,
        out_dir=".",
        out_fname_suffix=".cvg_comparison.csv",
        processes=0,
        cov_file_pattern=".qc.coverage.txt",
        column_dtypes=COV_COLUMN_DTYPES,
        metric="normalized_coverage",
    ):

        # store arguments
        try:
            self.file = os.path.abspath(file)
        except Exception:
            self.file = file
        try:
            self.batch = os.path.abspath(batch)
        except Exception:
            self.batch = batch
        try:
            self.intervals = os.path.abspath(intervals)
        except Exception:
            self.intervals = intervals
        self.dup_threshold = dup_threshold
        self.out_dir = os.path.abspath(out_dir)
        self.out_fname_suffix = out_fname_suffix
        self.processes = processes
        # use max cpu count available as number of processes if value not given
        if self.processes == 0:
            self.processes = mp.cpu_count()
        self.cov_file_pattern = cov_file_pattern
        self.column_dtypes = column_dtypes
        self.metric = metric

        # process a single file
        if self.file:
            # check and process the file
            self._process_file(
                file=self.file, intervals=self.intervals, out_dir=self.out_dir
            )

        # process a batch of files
        elif self.batch:
            # find the coverage files
            files = self._find_coverage_files(
                batch_dir=batch, cov_file_pattern=self.cov_file_pattern
            )
            # process the files in paralllel
            pool = mp.Pool(self.processes)
            # instance of class is passed to map (uses __call__
            # to call _process_file) for compatibility with python2.7
            results = pool.map(self, [file for file in files])
            pool.close()

        # print error if neither file or batch given
        else:
            sys.stderr.write(
                "[{0}] Error: File or batch folder not specified!".format(
                    datetime.now()
                )
            )
            sys.exit()

    @classmethod
    def _find_coverage_files(cls, batch_dir, cov_file_pattern):
        """ Method to find coverage files in a directory.
        :param: batch_dir: Path to batch directory
        :param: cov_file_pattern: Filename suffix to identify coverage files
        :returns: A list of coverage files found in given directory
        """
        print(
            "[{0}] Finding coverage files in {1}".format(
                datetime.now(), batch_dir
            )
        )

        try:
            coverage_files = glob.glob(
                os.path.join(batch_dir, "**", "*{0}".format(cov_file_pattern)),
                recursive=True,
            )
        # handle when glob recursive not available in older python versions
        except TypeError:
            coverage_files = []
            for root, dirnames, filenames in os.walk(batch_dir):
                for filename in fnmatch.filter(
                        filenames, "*{0}".format(cov_file_pattern)):
                    coverage_files.append(os.path.join(root, filename))
        except Exception as e:
            coverage_files = []
        return coverage_files

    @classmethod
    def _import_coverage(cls, file, column_dtypes, sep="\t"):
        """ Method to read output file from coverage file into a dataframe.
        :param: file: Path to coverage file
        :param: column_dtypes: Dictionary with column name and datatype for
                            the given coverage file type
        :param: sep: seperate used in coverage file (default is tab)
        :returns: reader iterable containing Pandas dataframe.
        """
        reader = pd.read_csv(
            file,
            sep=sep,
            header=0,
            skipinitialspace=True,
            na_values="-",
            dtype=column_dtypes,
            chunksize=10000,
        )

        return reader

    @classmethod
    def _import_intervals(cls, intervals, dup_threshold):
        """ Method to process interval json file and store as a dictionary
        :param: file: Path to interval json
        :param: dup_threshold: Threshold used to identify dup/amp regions
        :returns: Dictionary
        """
        # open json
        with open(intervals) as json_file:
            intervals = json.load(json_file)

        # validation json contents
        for interval in intervals:
            [interval[k] for k in ["region1", "region2"]]
            # add on default dup threshold if it has not been defined
            try:
                if interval["dup_threshold"] == "":
                    interval["dup_threshold"] = dup_threshold
            except KeyError:
                interval["dup_threshold"] = dup_threshold

        return intervals

    @classmethod
    def _filter_df(cls, df, filters):
        """ Filter df.
        :param: df: Pandas dataframe
        :parma: filter: Filter values as a dictionary
        :returns: filtered Pandas dataframe
        """
        query = " & ".join(
            ['{} == "{}"'.format(k, v) for k, v in filters.items()]
        )
        filtered_df = df.query(query)

        return filtered_df

    @classmethod
    def _get_sample_id(cls, file, cov_file_pattern):
        """ Method to return sample ID from file name.
        :param: file: Input file name
        :param: cov_file_pattern: Coverage file suffix to remove from filename
        :returns: sample id string
        """
        sample_id = (
            os.path.basename(file).split(cov_file_pattern)[0].replace("-", "_")
        )

        return sample_id

    def _compare_metric_at_intervals(self, cov_chunks, pair, metric):
        """ Method to compare normalised coverage at given intervals and score
        with a theshold value.
        :param: cov_chunks: List of Dataframes containing coverage info
        :param: pair: Dictionary containing info of intervals to be compared
        :param: metric: Column name in the df corresponding to the metric that
                        needs to be compared.
        :returns: Pandas dataframe
        """
        # logging
        print(
            "[{0}] Comparing {1} at {2} with {3}.".format(
                datetime.now(),
                metric,
                ",".join([str(i) for i in pair["region1"].values()]),
                ",".join([str(i) for i in pair["region2"].values()]),
            )
        )
        # iterate over each df chunk
        chunk_list = []
        for df in cov_chunks:
            for region in ["region1", "region2"]:
                # filter df for region of interest
                filt = self._filter_df(
                    df=df,
                    filters={
                        key: value
                        for key, value in pair[region].items()
                        if not key == "name"
                    },
                )
                if filt.shape[0] == 1:
                    # rename the interval for clarity
                    filt["name"] = pair[region]["name"]
                    chunk_list.append(filt)
                    # add on sample id column
                    filt["sample_id"] = self.sample_id

        # concatent the list or dataframe rows into a dataframe
        intv_comp = pd.concat(chunk_list)
        # convert from long to wide
        intv_comp = intv_comp.pivot(
            index="sample_id", columns="name", values=metric
        )
        # calc fold change between the two values
        intv_comp["fold_change"] = round(
            (
                intv_comp[pair["region1"]["name"]]
                / intv_comp[pair["region2"]["name"]]
            ),
            9,
        )
        # take a log of the fold change
        intv_comp["log2_fold_change"] = round(
            np.log2(intv_comp["fold_change"]), 9
        )

        # is the score above given theshold?
        intv_comp["above_cut_off"] = "FALSE"
        intv_comp.loc[
            intv_comp["fold_change"] > pair["dup_threshold"], "above_cut_off",
        ] = "TRUE"

        # reset index and sort
        intv_comp = intv_comp.reset_index().sort_values(
            by=["fold_change", "sample_id"], ascending=False
        )
        # order the columns
        intv_comp = intv_comp[
            [
                "sample_id",
                pair["region1"]["name"],
                pair["region2"]["name"],
                "fold_change",
                "log2_fold_change",
                "above_cut_off",
            ]
        ]

        return intv_comp

    def _process_file(self, file, intervals, out_dir):
        """ Method to process the coverage file to identify potential
        breakpoints.
        :param: file: Path to coverage file
        :param: intervals: Path of interval json file
        :param: out_dir: Path to output fir
        :returns: Writes a list of coverage comparisons and returns
                True if file processed, othwise False.
        """
        print(
            "[{0}] Processing {1}".format(
                datetime.now(), os.path.basename(file)
            )
        )
        # add to self so can be used elsewhere
        self.file = file
        # get sample
        self.sample_id = self._get_sample_id(
            file=self.file, cov_file_pattern=self.cov_file_pattern
        )
        # parse interval file to extract which region we want to process
        try:
            intervals = self._import_intervals(
                intervals=intervals, dup_threshold=self.dup_threshold
            )
        except Exception as e:
            sys.stderr.write(str(e))
            sys.stderr.write(
                "[{0}] Error: {1} interval file wrong format! Exiting...!"
                .format(
                    datetime.now(), os.path.basename(intervals)
                )
            )
            sys.exit()
        # parse the coverage file into chunks - helps with memory requirements
        # for v. large files
        try:
            cov_chunks = self._import_coverage(
                file=file, column_dtypes=self.column_dtypes
            )
        except Exception as e:
            sys.stderr.write("[{0}] Error: {1}".format(datetime.now(), str(e)))

        # for each interval pair find and compare the interval
        try:
            out_list = []
            for pair in intervals:
                out = self._compare_metric_at_intervals(
                    cov_chunks, pair, metric=self.metric
                )

                out_list.append(out)
            # concatent into single dataframe
            out_df = pd.concat(out_list, axis=0, ignore_index=True)
            # create filename
            fname = os.path.join(
                out_dir, "{0}{1}".format(self.sample_id, self.out_fname_suffix)
            )
            self.fname = fname
            # write output to file
            out_df.to_csv(fname, index=False)
            # logging
            print("[{0}] Results written to {1}".format(datetime.now(), fname))
        except Exception as e:
            sys.stderr.write("[{0}] Error: {1}".format(datetime.now(), str(e)))
            sys.stderr.write(
                "[{0}] Error: {1} tsv malformed/missing intervals".format(
                    datetime.now(), os.path.basename(file)
                )
            )

    def __call__(self, file):
        return self._process_file(
            file=file, intervals=self.intervals, out_dir=self.out_dir
        )


if __name__ == "__main__":
    # parse the command line arguments
    args = tdh_argument_parser(sys.argv[1:])
    # process the coverage file(s)
    TDHunter(
        file=args.file,
        batch=args.batch,
        intervals=args.intervals,
        dup_threshold=args.dup_threshold,
        out_dir=args.out_dir,
        out_fname_suffix=args.out_fname_suffix,
        processes=args.processes,
        cov_file_pattern=args.cov_file_pattern,
        metric=args.metric,
    )
